File: frontend/types/feature.ts
export interface PlatformFeature {
  id: string;
  key: string;
  name: string;
  is_premium: boolean;
  enabled_for_all_until?: string | null;
  experiment_group?: string | null;
  rollout_percentage?: number;
  feature_usage_limit?: number | null;
  current_price?: number | null;
  suggested_price_min?: number | null;
  suggested_price_max?: number | null;
  last_price_analysis_at?: string | null;
  has_access?: boolean | null;
  access_reason?: string | null;
  remaining_trial_uses?: number | null;
  trial_usage_count?: number | null;
  effective_trial_limit?: number | null;
  rollout_eligible?: boolean | null;
  experiment_variant?: string | null;
  user_segment?: string | null;
  recommended_prompt_intensity?: "default" | "aggressive" | null;
  created_at: string;
}

export interface UpdatePlatformFeaturePayload {
  is_premium?: boolean;
  enabled_for_all_until?: string | null;
  experiment_group?: string | null;
  rollout_percentage?: number;
  feature_usage_limit?: number | null;
  current_price?: number | null;
}

File: backend/models/feature.py
"""
Feature flag catalog for premium entitlements.
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from decimal import Decimal

from sqlalchemy import Boolean, DateTime, Integer, Numeric, String
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from database.base import Base


class Feature(Base):
    """Product feature that can be premium-gated or temporarily opened to everyone."""

    __tablename__ = "features"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
    )
    key: Mapped[str] = mapped_column(
        String(100),
        unique=True,
        nullable=False,
        index=True,
    )
    name: Mapped[str] = mapped_column(
        String(255),
        nullable=False,
    )
    is_premium: Mapped[bool] = mapped_column(
        Boolean,
        default=True,
        nullable=False,
        server_default="true",
    )
    enabled_for_all_until: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )
    experiment_group: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
        index=True,
    )
    rollout_percentage: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=0,
        server_default="0",
    )
    feature_usage_limit: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
    )
    current_price: Mapped[Decimal | None] = mapped_column(
        Numeric(10, 2),
        nullable=True,
    )
    suggested_price_min: Mapped[Decimal | None] = mapped_column(
        Numeric(10, 2),
        nullable=True,
    )
    suggested_price_max: Mapped[Decimal | None] = mapped_column(
        Numeric(10, 2),
        nullable=True,
    )
    last_price_analysis_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    def __repr__(self) -> str:
        return (
            f"<Feature(key={self.key!r}, premium={self.is_premium}, "
            f"rollout={self.rollout_percentage}, usage_limit={self.feature_usage_limit}, "
            f"current_price={self.current_price})>"
        )
